### PR TITLE
Change "inCall" state to contain bit flags (#911).

### LINKED ISSUE #911: Use bit flags instead of boolean for "inCall" flag.
Currently the flag `inCall` is a boolean field that is `true` when a participant is in the call and `false` when in the room (but not the call yet).

I suggest to change this to an integer so different flags can be encoded into it:
- in call
- with audio
- with video
- with screensharing
- etc.

This will allow showing different icons in the participant list or hiding the video panel if no video is published and will also help with the MCU/phone integration.

@Ivansss @nickvergessen What do you think? I could start working on a PR.

### PR DESCRIPTION
**Pull request by fancycode**
The flags encode if a user is in the call and whether he is publishing audio and/or video. If a participant is not publishing streams, it will not be called/call other participants.

Open issues:
- [x] UI improvements: currently the UI switches to video mode for everybody, even for the single publisher in a 1-1 call (-> defer to a follow-up PR).
- [x] Collapse the database migration to a single step if possible. I couldn't find a way to alter the column from boolean to int with a single commant (keep as-is).
- [x] Add `in-call-flags` capability

In the future, the flags could be used to show different icons in the participant list. Also they could be changed when a participant mutes his audio/video and for example the video panel could be completely hiddeb in this case.

@Ivansss maybe you can help with the UI improvements?

Fix #911

### CODE REVIEW

**Comment by Ivansss** [comment]
> If a participant is not publishing streams, it will not be called/call other participants.

We would like to keep the possibility of joining a call without mic neither camera (just as a listener).
Also it would be nice to have another flag for screensharing.
We should keep in mind that people without camera and mic would be able to share their screen.

**Comment by fancycode** [comment]
> We would like to keep the possibility of joining a call without mic neither camera (just as a listener).

Sure, that is possible. However these participants don't need a PeerConnection with other people that also don't have audio and video.

> Also it would be nice to have another flag for screensharing.

I would do that in a separate PR.

> We should keep in mind that people without camera and mic would be able to share their screen.

This is working as screensharing is using a separate PeerConnection anyway, or am I missing something?

**Comment by Ivansss** [comment]
> Sure, that is possible. However these participants don't need a PeerConnection with other people that also don't have audio and video.

That's right. But we should point somehow that these participants are in the call (not just in the room) even they are not streaming anything. Then participants that are streaming something send them offers.

**Comment by fancycode** [comment]
> That's right. But we should point somehow that these participants are in the call (not just in the room) even they are not streaming anything. Then participants that are streaming something send them offers.

The participants that are in the room but don't publish audio/video also have a bit set in the flags:
https://github.com/nextcloud/spreed/pull/914/files#diff-726e60e30a9cf2e43e50f8bbbff90908R111

**Comment by mario** [comment]
@fancycode why do we open a separate connection for screensharing? That's always been my question. Isn't it more natural to just add another video track?

**Comment by mario** [comment]
@Ivansss shouldn't inCall flag represent the fact you're in the call? (NOT IN THE ROOM, but actually in the CALL)

**Comment by Ivansss** [comment]
> The participants that are in the room but don't publish audio/video also have a bit set in the flags:
> https://github.com/nextcloud/spreed/pull/914/files#diff-726e60e30a9cf2e43e50f8bbbff90908R111

Oh! My mistake then. Could we rename that flag to `FLAG_IN_CALL`?

**Comment by Ivansss** [comment]
Yeah, I will add a new capability. Like `in-call-flags`

File: lib/Room.php [hunk 1]
```diff
@@ -597,12 +603,16 @@ public function joinRoomGuest($password, $passedPasswordProtection = false) {
 
 	/**
 	 * @param string $sessionId
-	 * @param bool $active
+	 * @param int $flags
 	 */
-	public function changeInCall($sessionId, $active) {
-		if ($active) {
+	public function changeInCall($sessionId, $flags) {
+		if ($flags === false) {
```
Comment (nickvergessen) [inline lib/Room.php]: This should simply not be allowed anymore

File: lib/Controller/CallController.php [hunk 2]
```diff
@@ -131,9 +132,10 @@ public function getPeersForCall($token) {
 	 * @UseSession
 	 *
 	 * @param string $token
+	 * @param int $flags
 	 * @return DataResponse
 	 */
-	public function joinCall($token) {
+	public function joinCall($token, $flags) {
```
Comment (nickvergessen) [inline lib/Controller/CallController.php]: So would a user reconnect when enabling/disabling the camera/audio? Or would this just not change at all?

Comment (fancycode) [reply lib/Controller/CallController.php]: For now the flags are only evaluated once as result to the `getUserMedia` when joining the call. Using that, the UI on the receiver side could look differently for users that don't have / allow access to their camera/microphone.

Updating the flags when the audio/video is enabled/disabled was not part of the scope of this PR (I would rather do that in a separate PR).

**Comment by nickvergessen** [comment]
I rebased this on master to resolve the conflicts and added a const for the magic 0 aswell.

Anything left todo @fancycode

**Comment by nickvergessen** [comment]
This is stable14+ only.

**Pull request merged**